libr/debug/p/debug_evm.c

```diff
@@ -0,0 +1,287 @@
+
+#include <r_asm.h>
+#include <r_debug.h>
+
+typedef struct {
+	size_t curr_instruction;
+
+	ut64 *breakpoints;
+	size_t breakpoints_capacity;
+	size_t breakpoints_length;
+
+} RIOEvmState;
+
+typedef struct {
+	uint8_t	depth;
+	uint8_t error;
+	unsigned pc;
+	unsigned gas;
+	unsigned gas_cost;
+
+	uint8_t *memory;
+	size_t memlength;
+
+	uint8_t *stack;
+	size_t stack_length;
+	char *op;
+} RIOEvmOp;
+
+typedef struct {
+	void *curl;
+	char *host;
+	int port;
+	char *tx;
+	char *tx_full;
+	char *tx_to;
+	char *tx_from;
+	char *to_code_resp;
+	char *to_code;
+
+	uint8_t *code;
+	size_t code_size;
+
+	char *response;
+	size_t curr_resp_size;
+
+	RIOEvmOp *ops;
+	size_t ops_size;
+
+    size_t curr_op;
+} RIOEvm;
+
+static RIOEvm *rio = NULL;
+static RIOEvmState *rios = NULL;
+
+static int r_debug_evm_step(RDebug *dbg) {
+    if (!rio->ops_size || rios->curr_instruction >= rio->ops_size) {
+        return false;
+    }
+
+	rios->curr_instruction++;
+    rio->curr_op = rios->curr_instruction;
+
+	return true;
+}
+
+static RList* r_debug_evm_threads(RDebug *dbg, int pid) {
+	return NULL;
+}
+
+static int r_debug_evm_reg_read(RDebug *dbg, int type, ut8 *buf, int size) {
+	if (!rios || !rio || !rio->ops_size) {
+		return 0;
+	}
+	((uint16_t*)(buf))[0] = (uint16_t)rio->ops[rios->curr_instruction].pc;
+	((uint16_t*)(buf))[1] = (uint16_t)0x8fff;
+	return 4;
+}
+
+static RList *r_debug_evm_map_get(RDebug *dbg) {
+    RList *list;
+    RDebugMap *map;
+
+    if (rio && !rio->ops_size) {
+        return NULL;
+    }
+
+    list = r_list_new();
+
+    map = r_debug_map_new ("code", 0x0000, 0x7FFF, r_str_rwx("rwx"), 0);
+    r_list_append (list, map);
+
+    map = r_debug_map_new ("stack", 0x8FFF, 0xFFFF, r_str_rwx("rwx"), 0);
+    r_list_append (list, map);
+
+	map = r_debug_map_new ("memory", 0x10000, 0x1FFFF, r_str_rwx("rwx"), 0);
+	r_list_append (list, map);
+
+	return list;
+}
+
+static int r_debug_evm_reg_write(RDebug *dbg, int type, const ut8 *buf, int size) {
```

Comment:
Either it's bool or int, it's a bad idea to return `false` in case of the `int` function.